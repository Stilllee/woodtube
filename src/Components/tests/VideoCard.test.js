import { MemoryRouter, Route, Routes } from "react-router-dom";

import React from "react";
import VideoCard from "../VideoCard";
import { formatAgo } from "../../util/date";
import { render } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import { useLocation } from "react-router-dom";
import userEvent from "@testing-library/user-event";

describe("VideoCard", () => {
  const video = {
    id: 1,
    snippet: {
      title: "제목",
      channelId: "1",
      channelTitle: "채널명",
      publishedAt: new Date(),
      thumbnails: {
        medium: {
          url: "http://example.com/image.jpg",
        },
      },
    },
  };
  const { title, channelId, channelTitle, publishedAt, thumbnails } =
    video.snippet;

  it("비디오 항목을 렌더링", () => {
    render(
      <MemoryRouter>
        <VideoCard video={video} />
      </MemoryRouter>
    );

    const image = screen.getByRole("img");
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatAgo(publishedAt, "ko"))).toBeInTheDocument();
  });

  it("클릭하면 video state와 함께 비디오 상세 페이지로 이동", async () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }

    render(
      <MemoryRouter initialEntries={["/"]}>
        <Routes>
          <Route path="/" element={<VideoCard video={video} />} />
          <Route
            path={`/videos/watch/${video.id}`}
            element={<LocationStateDisplay />}
          />
        </Routes>
      </MemoryRouter>
    );

    const card = screen.getByRole("listitem");
    await userEvent.click(card);

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
