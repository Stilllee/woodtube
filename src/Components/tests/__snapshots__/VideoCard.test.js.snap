// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard 그리드 type을 올바르게 렌더링 1`] = `
<DocumentFragment>
  <li
    class=" cursor-pointer hover:scale-105 transition-transform rounded-xl"
  >
    <img
      alt="제목"
      class="w-full rounded-xl"
      src="http://example.com/image.jpg"
    />
    <div>
      <p
        class="my-2 font-semibold line-clamp-2"
        title="제목"
      >
        제목
      </p>
      <p
        class="text-sm opacity-80"
      >
        채널명
      </p>
      <p
        class="text-sm opacity-80"
      >
        방금
      </p>
    </div>
  </li>
</DocumentFragment>
`;

exports[`VideoCard 리스트 type을 올바르게 렌더링 1`] = `
<DocumentFragment>
  <li
    class="flex gap-1 cursor-pointer hover:scale-105 transition-transform rounded-xl"
  >
    <img
      alt="제목"
      class="w-60 mr-2 rounded-xl"
      src="http://example.com/image.jpg"
    />
    <div>
      <p
        class="my-2 font-semibold line-clamp-2"
        title="제목"
      >
        제목
      </p>
      <p
        class="text-sm opacity-80"
      >
        채널명
      </p>
      <p
        class="text-sm opacity-80"
      >
        방금
      </p>
    </div>
  </li>
</DocumentFragment>
`;
